src/sentry_core.c

```diff
@@ -633,3 +656,94 @@ sentry_set_level(sentry_level_t level)
         scope->level = level;
     }
 }
+
+bool
+sentry__should_sample_transaction(sentry_value_t tx_cxt)
+{
+    sentry_value_t context_setting = sentry_value_get_by_key(tx_cxt, "sampled");
+    if (!sentry_value_is_null(context_setting)) {
+        return sentry_value_is_true(context_setting);
+    }
+
+    bool sample = false;
+    SENTRY_WITH_OPTIONS (options) {
+        sample = options->traces_sample_rate == 1.0;
+        // TODO: roll rng if sample rate >0 and <1.0
+        // TODO: run through traces sampler function if rate is unavailable
+    }
+    return sample;
+}
+
+sentry_value_t
+sentry_start_transaction(sentry_value_t tx_cxt)
+{
+    // TODO: stuff transaction into the scope
+    bool should_sample = sentry__should_sample_transaction(tx_cxt);
+
+    sentry_value_t tx = sentry_value_new_event();
+
+    // TODO: does it matter that this is sampled on creation vs on send? can we
+    // just sample on send like other events?
+    sentry_value_set_by_key(
+        tx, "sampled", sentry_value_new_bool(should_sample));
+
+    sentry_value_set_by_key(
+        tx, "transaction", sentry_value_get_by_key_owned(tx_cxt, "name"));
+    sentry_value_set_by_key(tx, "start_timestamp",
+        sentry__value_new_string_owned(
+            sentry__msec_time_to_iso8601(sentry__msec_time())));
+
+    // eventually goes into "contexts": { "trace": trace_context }
+    sentry_value_t trace_context = sentry_value_new_object();
+
+    sentry_uuid_t trace_id = sentry_uuid_new_v4();
+    sentry_value_set_by_key(
+        trace_context, "trace_id", sentry__value_new_internal_uuid(&trace_id));
+
+    sentry_uuid_t span_id = sentry_uuid_new_v4();
+    sentry_value_set_by_key(
+        trace_context, "span_id", sentry__value_new_span_uuid(&span_id));
+
+    sentry_value_set_by_key(
+        trace_context, "op", sentry_value_get_by_key_owned(tx_cxt, "op"));
+
+    sentry_value_set_by_key(
+        trace_context, "status", sentry_value_new_string("ok"));
+
+    // TODO: add description to transaction context and
+    // trace_context,
+    // TODO: integrate with new sentry__span_get_trace_context
+    sentry_value_set_by_key(tx, "trace_context", trace_context);
+
+    sentry_value_decref(tx_cxt);
+
+    return tx;
+}
+
+void
+sentry_transaction_finish(sentry_value_t tx)
+{
+    sentry_value_set_by_key(tx, "type", sentry_value_new_string("transaction"));
```

Comment:
Why are we setting the type here?

Reply:
good question; if you're wondering why it's set in finish of all places, i'd have to say that there's no real rhyme or reason behind it besides the idea that the transactions _shouldn't_ be treated like a regular event (ie require checking of the type field) elsewhere in the code after their creation. that's likely a naive assumption given the lack of typing we have for transactions, though. now that i've written this out, i might just set this field during initialization to be safe.

if you're wondering why this field is getting set at all (and why it isn't being set for events), it's the discriminant in "events" (https://develop.sentry.dev/sdk/event-payloads/types/#eventtype). tl;dr SDKs send two types of events, error monitoring-related events and transactions. the former is the default and doesn't need an explicit `type`, the latter must have the `type` field set to transaction for it to be treated as such.